src/ring_theory/ideal/nc_jacobson_radical.lean

```diff
@@ -0,0 +1,251 @@
+let ⟨⟨_, _, hxa, hax⟩, rfl⟩ := h in ⟨⟨_, _, by rw [←unop_mul, ←unop_one, hax],
+  by rw [←unop_mul, ←unop_one, hxa]⟩, rfl⟩
+
+lemma is_unit_op_iff_is_unit {a : M} : is_unit (op a) ↔ is_unit a :=
+⟨λ h, unop_op a ▸ h.unop, is_unit.op⟩
+
+lemma is_unit_unop_iff_is_unit {x : Mᵐᵒᵖ} : is_unit (unop x) ↔ is_unit x :=
+⟨λ h, op_unop x ▸ h.op, is_unit.unop⟩
+
+lemma is_unit_iff_has_two_sided_inv :
+  is_unit a ↔ ∃ x : M, a * x = 1 ∧ x * a = 1 :=
+⟨λ ⟨⟨a, x, hax, hxa⟩, rfl⟩, ⟨x, hax, hxa⟩, λ ⟨x, hax, hxa⟩, ⟨⟨a, x, hax, hxa⟩, rfl⟩⟩
+
+/-- An element `a : M` of a monoid has a left inverse
+if there is some `x : M` satisfying `x * a = 1`. -/
+def has_left_inv (a : M) : Prop := ∃ x : M, x * a = 1
+
+/-- An element `a : M` of a monoid has a right inverse
+if there is some `x : M` satisfying `a * x = 1`. -/
+def has_right_inv (a : M) : Prop := ∃ x : M, a * x = 1
+
+/-- An element of a monoid is a unit
+if and only if it has both a left inverse and a right inverse. -/
+theorem is_unit_iff_has_left_inv_right_inv :
+  is_unit a ↔ has_left_inv a ∧ has_right_inv a :=
+⟨λ h, ⟨h.exists_left_inv, h.exists_right_inv⟩,
+  λ ⟨⟨x, hxa⟩, ⟨y, hay⟩⟩, ⟨⟨a, x, by { convert hay,
+  rw [←mul_one x, ←hay, ←mul_assoc, hxa, one_mul] }, hxa⟩, rfl⟩⟩
+
+/-- An element of a monoid is a unit
+if it has a left inverse which also has a left inverse. -/
+theorem is_unit_of_has_left_inv_of_has_left_inv
+  (h₁ : b * a = 1) (h₂ : has_left_inv b) : is_unit a :=
+let ⟨c, hcb⟩ := h₂ in ⟨⟨a, b, by { convert hcb,
+  rw [←one_mul a, ←hcb, mul_assoc, h₁, mul_one] }, h₁⟩, rfl⟩
+
+end monoid
+
+section ring
+variables {R : Type u} {a : R} {b : R} [ring R]
+/-! ### Some lemmas on inverses in rings -/
+
+lemma has_left_inv.one_add_mul_swap
+  (h : has_left_inv (1 + a * b)) : has_left_inv (1 + b * a) :=
+begin
+  obtain ⟨u, hu⟩ := h,
+  existsi 1 - b * u * a,
+  calc (1 - b * u * a) * (1 + b * a)
+        = 1 + b * a - b * (u * (1 + a * b)) * a : by noncomm_ring
+    ... = 1 : by rw [hu, mul_one, add_sub_cancel],
+end
+
+lemma has_right_inv.one_add_mul_swap
+  (h : has_right_inv (1 + a * b)) : has_right_inv (1 + b * a) :=
+begin
+  obtain ⟨u, hu⟩ := h,
+  existsi 1 - b * u * a,
+  calc (1 + b * a) * (1 - b * u * a)
+        = 1 + b * a - b * ((1 + a * b ) * u) * a : by noncomm_ring
+    ... = 1 : by rw [hu, mul_one, add_sub_cancel],
+end
+
+lemma is_unit.one_add_mul_swap
+  (h : is_unit (1 + a * b)) : is_unit (1 + b * a) :=
+begin
+  rw is_unit_iff_has_left_inv_right_inv at *,
+  obtain ⟨h₁, h₂⟩ := h,
+  exact ⟨h₁.one_add_mul_swap, h₂.one_add_mul_swap⟩,
+end
+
+end ring
+
+namespace ideal
+section nc_jacobson_radical
+variables {R : Type u} [ring R]
+/-! ### Jacobson radical of a ring -/
+
+/--
+For a semiring `R`, `nc_jacobson R` is the Jacobson radical of `R`, that is,
+the intersection of all maximal left ideals of of `R`. Note that we use left ideals.
+-/
+def nc_jacobson (R : Type u) [semiring R] : ideal R :=
+Inf {I : ideal R | I.is_maximal }
```

Comment:
I think it would be bette rto just change `comm_ring` to `ring` on `ideal.jacobson` - you don't need to change all the proofs to work in that weaker setting, but it would at least allow you to reuse the definition.

Reply:
I've attempted that change in #13865 - I haven't done anything mathematically interesting there; so I think we'll still want lots of stuff from this (your) PR :)

Reply:
I think you mean `ring_theory.jacobson_ideal`. I have a feeling that importing it in this `nc_jacobson_radical` is not a good choice.
The first reason is that, as contrast to commutative case, in noncommutative setting, the notion of the Jacobson radical of a (left, right, two-sided) ideal is very rare (I haven't seen it). This is because it conflicts with more frequently used notion: the Jacobson radical of a module (Inf of all maximal submodules).
The second is that, actually the proofs of some theorems in `ring_theory.jacobson_ideal` (whose name contains `jacobson_bot`) could be simplified using this PR (I'm not sure whether such a thing is worth doing). If we do so, then I think a circular import occurs.

Reply:
@haruhisa-enomoto, I'm confused by this objection. Isn't your definition exactly a special case of the definition in `ring_theory.jacobson_ideal` (after Eric's generalisation)?

If anything, we should just further generalise the definition in `ring_theory.jacobson_ideal` so it applies to arbitrary modules, not just ideals. Is there actually a conflict?

Reply:
Yes it's a special case, but at least a confusion occurs if we simply generalize Jacobson radical to modules.
In `ring_theory.jacobson_ideal`, the Jacobson radical of an ideal `I` of `R` is the intersection of maximal submodules of `R` containing `I`, but when `I` is considered as a module, then the Jacobson radical of `I` should be the intersection of "maximal submodules of `I`" (so it's smaller than `I`).
Actually no conflicts occur by defining `jacobson N` of `N: submodule R M` as the intersection of maximal sub of `M` containing `N`, then define `module_jacobon M` of `M : module R` as `jacobson (⊥ : submodule R M)`.
Should I go in this way?

Reply:
Yes, I think you should go this way, but perhaps you don't even need to define `module_jacobson`.

I think the sequence should be:
* take the existing material for `jacobson I`, and splits it into the parts that work just for a module, and the parts that only work for an ideal.
* the parts that work for any submodule will presumably trivially generalise from comm_ring to ring
* the parts that only work for an ideal may require checking to see where commutativity is used

This could perhaps be done in three separate PRs, so it's easy to understand what happens at each stage? But as you prefer, if you want to do it. :-)

Reply:
`nc_jacobson` is a special case of `module_jacobson` right? Maybe call both of them `jacobson_radical`?

Reply:
@semorrison I see the way. So in this PR, I will
- Change `jacobson` in `ring_theory.jacobson_ideal` to include noncommutative (semi)ring
- Import `ring_theory.jacobson_ideal` in this `nc_jacobson_radical`, and use it to define `nc_jacobson R` for `R : (semi)ring`.
Is it OK?
And how about duplications, e.g. (mem_jacobson_bot)[https://github.com/leanprover-community/mathlib/blob/922717e8c7fbbd2230265b576a4504b9be28039d/src/ring_theory/jacobson_ideal.lean#L212] in `ring_theory.jacobson_ideal` is clearly contained in this `nc_jacobson_radical`.

Reply:
@alreadydone I want call them `jacobson_radical`, but then we have two `jacobson` and `jacobson_radical`, and probably usual commutative people will want to call `jacobson` of an ideal the Jacobson radical of an ideal, which may be confusing...

Reply:
Maybe I'm confused here (this is beyond my ring-theory pay grade by now), but why does `nc_jacobson` need a separate definition at all? Isn't it just going to be `jacobson ⊥`?

Reply:
> So in this PR, I will Change jacobson in ring_theory.jacobson_ideal to include noncommutative (semi)ring

You should wait for #13865 so as not to repeat this work.

Reply:
OK I will wait for it.

Reply:
It's now merged; you can merge master into your branch to get the changes

Reply:
Thank you! I've done it and changed the definition of `nc_jacobson`.